Automattic/newspack-popups
# feat: allow admins to define merge fields for donors

*|### All Submissions:






### Changes proposed in this Pull Request:



Refactors the Campaigns Settings page to use the new `PluginSettings` component (requires https://github.com/Automattic/newspack-plugin/pull/1450).

Creates a new setting option to define a comma-delimited list of strings to look for when checking Mailchimp merge fields (instead of hardcoding `DONAT` as the only string to check for). Closes #684.

### How to test the changes in this Pull Request:

#### New Settings UI

1. Follow testing instructions in https://github.com/Automattic/newspack-plugin/pull/1450 to test the wizard UI experience.
2. Deactivate both `newspack-plugin` and `newspack-blocks` to enable the "fallback" settings UI.
3. You should now see a **Prompts > Settings** link in the WP admin dashboard sidebar menu. Click it.
4. Confirm that all options are available here, and that you now have to click "Save" to save your settings (previously the settings updated when an option value changes, but that doesn't work as well for text inputs like the new Mailchimp Donor Merge Fields option).
5. Change various options, save, and refresh the page to confirm that your settings are saved.

<img width="535" alt="Screen Shot 2022-01-21 at 5 02 54 PM" src="https://user-images.githubusercontent.com/2230142/150615150-854207ce-13f9-4889-918e-3487f6b9f22d.png">

#### Donor merge fields

This one's a bit involved to test. Before getting started, make sure you have the following prerequisites setup:

* A Mailchimp account with a list and at least one subscriber whose email address you have access to (DM me if you need access to a Mailchimp account to test with)
* A test site with Newspack Newsletters connected to and setup to send to said Mailchimp list
* A Campaigns segment set to "Subscribers" only, with at least one prompt to show to subscribers
* A Campaigns segment at the highest priority, set to "Donors" only, with at least one prompt to show to donors

Once you have the above, to test:

1. Make sure Newspack Plugin is checked out to https://github.com/Automattic/newspack-plugin/pull/1450.
2. Go to **Newspack > Campaigns > Settings**. Observe the new option to define Mailchimp merge tags, and that it defaults to `DONAT` for backwards compatibility:
 
<img width="504" alt="Screen Shot 2022-01-21 at 1 53 26 PM" src="https://user-images.githubusercontent.com/2230142/150598704-2dbd7aff-400f-4336-b25f-6b74ba711d5e.png">

4. Update the field to something else (e.g. `HUB_` or `DONAT,HUB_`) and save. Refresh the page and confirm that your setting is saved.
5. In Mailchimp, go to **Audience > All contacts** and expand the **Settings** menu. Select **Audience fields and \*|MERGE|\* tags**.

<img width="282" alt="Screen Shot 2022-01-21 at 1 56 19 PM" src="https://user-images.githubusercontent.com/2230142/150598983-0b4e0cee-f7a8-440c-b7fd-a44f0d6d4e1a.png">

6. Click **Add a Field** at the bottom of the page. Choose "Text" as the field type. A new Untitled field will appear at the bottom of the UI.
7. Set the label to something identifiable and (important) the "Put this tag in your content" field to something that matches a string in the **Mailchimp merge fields** option you set in Campaigns settings, e.g.:

<img width="1241" alt="Screen Shot 2022-01-21 at 1 59 20 PM" src="https://user-images.githubusercontent.com/2230142/150599335-2f5c4221-cd8a-4a64-bc2e-f06e136f717f.png">

(Note that in the above screenshot, the merge field label is `HUB_SUPPORTER` and the merge field name is `HUB_DONOR`—the latter is what's used to match to our plugin option.)

7. Save changes and go back to your All Contacts list view. Click on the name of a contact for which you have inbox access (so you can get the email when it's sent). Click "edit" in the **Profile Information** section and set the value of your new merge field to `true`. Save and exit.

<img width="452" alt="Screen Shot 2022-01-21 at 4 43 23 PM" src="https://user-images.githubusercontent.com/2230142/150613911-0ff7e326-abb5-4e9e-9975-236114e24f79.png">

8. Back in your site, create a newsletter with a link to a post on your site, and send it to your Mailchimp list.
9. Find the newsletter email and copy the link to the post. It should look something like `https://blog.us10.list-manage.com/track/click?u=ae81335f8566492521d60e635&id=82da4b0fff&e=754835379c`.
10. Visit the link in a new incognito session. (Optional: add the `?newspack-campaigns-debug` flag to the URL to be able to view debug info on the request response in dev tools.)
11. On the first post you view, confirm that you see the prompt for subscribers (donation status is processed on that first page load, so you won't be considered a donor until subsequent page views).
12. Navigate to another post and confirm that you now see the prompt for donors.

### Other information:
